keps/0013-external-specs.md

```diff
@@ -0,0 +1,162 @@
+---
+kep-number: 0013
+title: external-specs
+authors:
+  - "@runyontr"
+owners:
+  - @runyontr
+  - "@gerred"
+creation-date: 2019-06-18
+last-updated: 2019-06-18
+status: provisional
+---
+
+# External Application Specs
+
+## Table of Contents
+
+- [Table of Contents](#table-of-contents)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories [optional]](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+## Summary
+
+Organizations and companies already spend time building, updating and debugging specs for deploying their applications. Asking them to spend additional time to maintain a KUDO spec may prevent adoption of the platform. Improving KUDO to be able to consume more than just a KUDO Spec defined in [KEP-0009](keps/0009-operator-toolkit.md) would allow for oragnizations to provide access to their applications to the KUDO community.
+
+## Motivation
+
+While being able to package our own application definition via [KEP0009](keps/0009-operator-toolkit.md), being able to leverage other application package formats for customization and extension would provide access to a trove of KUDO Frameworks without any additional work required by FrameworkVendors. There are
+
+- Frameworks
+- Helm
+- CNAB
+- OperatorHub
+
+### Goals
+
+- Run non-KUDO defined applications as Frameworks
+
+### Non-Goals
+
+- (currently) Manage repos of non-KUDO applications (e.g. Helm Charts)
+- Convert non-KUDO applications into KUDO format
+- Create a plugin system to upload new templating engines. Must be present in core KUDO codebase (for now)
+- Adjusting the capabilities of Parameters to support non-string values
+
+## Proposal
+
+### Engine Interface
+
+The core functionality required by KUDO is the ability to turn
+
+```go
+type StepArgument struct{
```

Comment:
is it maybe `ExecutionStep`? 🤔 `ActiveStep`?

Reply:
This object is all that gets provided to the templating engine to be able to render the Kubernetes objects.  It should contain all the info needed.  It contains (currently planned) the FrameworkVersion, so the templates are available, as well as the parameter values in the Instance that is being rendered.

Some frameworks use other data provided by KUDO as part of their rendering, and this will also contain that data, though maybe it makes more sense to embed that in the parameter map.  For example `STEP_NAME`, `PLAN_NAME`, etc.